device-connectors/src/testflinger_device_connectors/devices/zapper_iot/__init__.py

```diff
@@ -37,16 +37,41 @@ def _validate_configuration(
         Validate the job config and data and prepare the arguments
         for the Zapper `provision` API.
         """
-        try:
-            provision_plan = self.job_data["provision_data"]["provision_plan"]
-            validate_provision_plan(provision_plan)
-        except KeyError as e:
-            raise ProvisioningError from e
 
+        username = self.job_data.get("test_data", {}).get(
+            "test_username", "ubuntu"
+        )
+        password = self.job_data.get("test_data", {}).get(
+            "test_password", "ubuntu"
+        )
+
+        provisioning_data = {
+            "username": username,
+            "password": password,
+            "preset": self.job_data["provision_data"].get("preset"),
+            "reboot_script": self.config["reboot_script"],
+        }
+
+        provision_plan = self.job_data["provision_data"].get("provision_plan")
+        if provision_plan:
+
+            try:
+                validate_provision_plan(provision_plan)
+
+                # Make sure the user created at provision time is
+                # the same used during the test phase.
+                provision_plan["config"]["username"] = username
+                provision_plan["config"]["password"] = password
```

Comment:
So, once this merged, we have to put our username and password to test_data in job yaml right?

Reply:
yes, exactly. This is because we need to make sure the same username/password are used in both provisioning *and* test phases.